src/BinSkim.Rules/PERules/BA6002.EliminateDuplicateStrings.cs

```diff
@@ -0,0 +1,147 @@
+﻿/// <summary>
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Reflection.PortableExecutable;
+
+using Microsoft.CodeAnalysis.BinaryParsers;
+using Microsoft.CodeAnalysis.BinaryParsers.PortableExecutable;
+using Microsoft.CodeAnalysis.BinaryParsers.ProgramDatabase;
+using Microsoft.CodeAnalysis.IL.Sdk;
+using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.Driver;
+
+namespace Microsoft.CodeAnalysis.IL.Rules
+{
+    [Export(typeof(ReportingDescriptor))]
+    [Export(typeof(Skimmer<BinaryAnalyzerContext>))]
+    public class EliminateDuplicateStrings : WindowsBinaryAndPdbSkimmerBase
+    {
+        /// <summary>
+        /// BA6002
+        /// </summary>
+        public override string Id => RuleIds.EliminateDuplicateStrings;
+
+        /// <summary>
+        /// The /GF compiler option, also known as Eliminate Duplicate Strings or String Pooling, will combine identical strings
+        /// in a program to a single readonly copy. This can significantly reduce binary size for programs with many string resources.
+        /// </summary>
+
+        public override MultiformatMessageString FullDescription =>
+            new MultiformatMessageString { Text = RuleResources.BA6002_EliminateDuplicateStrings_Description };
+
+        protected override IEnumerable<string> MessageResourceNames => new string[] {
+                    nameof(RuleResources.BA6002_Pass),
+                    nameof(RuleResources.BA6002_Warning),
+                    nameof(RuleResources.NotApplicable_InvalidMetadata)
+                };
+
+        private const string AnalyzerName = RuleIds.EliminateDuplicateStrings + "." + nameof(EliminateDuplicateStrings);
+
+        public override AnalysisApplicability CanAnalyzePE(PEBinary target, Sarif.PropertiesDictionary policy, out string reasonForNotAnalyzing)
```

Comment:
>CanAnalyzePE [](http://example.com/codeflow?start=46&length=12)

So, is CanAnalyzePE a shared implementation for  all the perf rules?

If so, maybe create a base class with a shared implementation.

Named something like

MsvcCompiledBinaryAndPdbBase
MicrosoftNativeCompilerGeneratedBinaryAndPdbBase

according to the rename choice you make.

Reply:
They are not all identical (although there is a lot of overlap).  The notable difference is that some rules only apply to Release builds whereas others apply to both Release and Debug.

For what it is worth I would much prefer a flat helper instead of any sort of class hierarchy for this sort of thing.  Some sort of `CanAnalyzePEForPerformance(bool allowDebug)` helper would do the trick.  I just need a reasonable place to put it.